File: cpp/trex_selector_methods/trex_screening/demo_trex_scr_03_mc_sim_correlated/demo_trex_scr_03_mc_sim_correlated.cpp
// ==============================================================================
// demo_trex_scr_03_mc_sim_correlated.cpp
// ==============================================================================
/**
 * @file demo_trex_scr_03_mc_sim_correlated.cpp
 *
 * @brief Screen-TRex on correlated predictor structures (AR1 and equi).
 *
 * @details
 *    1. Generate X with AR(1) or equicorrelation structure.
 *    2. Run Screen-TRex with the matching DA variant.
 *    3. Report FDP, TPP, estimated FDR, and estimated ρ.
 *
 *  Part 1 — SNR sweep: Ordinary vs. Bootstrap vs. DA-AR1 on AR(1) data.
 *  Part 2 — SNR sweep: Ordinary vs. Bootstrap vs. DA-EQUI on equicorrelated data.
 *  Part 3 — rho sweep: Ordinary vs. Bootstrap vs. DA-AR1 on AR(1) data.
 *  Part 4 — rho sweep: Ordinary vs. Bootstrap vs. DA-EQUI on equicorrelated data.
 *  Part 5 — rho sweep: Ordinary vs. Bootstrap vs. DA-BLOCK-EQUI on block-equicorr data.
 *
 *  All parts are Monte Carlo studies parallelised via OpenMP.
 */
// ==============================================================================

// std includes
#include <map>
#include <string>
#include <vector>

// Eigen includes
#include <Eigen/Dense>

// Screen-TRex demo simulation utilities
#include "trex_screening_simulation_utils.hpp"

using namespace scr_demo;



// ==============================================================================
// Part 1: SNR sweep — Ordinary vs. Bootstrap vs. DA-AR1 on AR(1) data
// ==============================================================================

/**
 * @brief SNR-sweep Monte Carlo: Ordinary vs. Bootstrap DA-AR1 Screen-TRex
 *        on AR(1)-correlated data.  Parallelised via OpenMP.
 *
 * @param num_MC    Number of Monte Carlo runs per (method, SNR) pair.
 * @param rho       AR(1) lag-1 correlation.
 * @param beta_val  Coefficient magnitude for active variables.
 */
void demo_DA_AR1_MonteCarlo(std::size_t num_MC,
                             double     rho      = 0.4,
                             double     beta_val = 1.0)
{
    std::cout.setf(std::ios::unitbuf);
    cdiag::print_section_header(
        "Part 1: DA-AR1 Screen-TRex Monte Carlo on AR(1) Data"
    );

    const Eigen::Index n = 300, p = 1000, p1 = 10;
    std::cout << "DGP: AR(1), n=" << n << ", p=" << p << ", p1=" << p1
              << ", rho=" << rho << "\n";

    const std::vector<double> snr_values =
        {0.01, 0.1, 0.2, 0.5, 0.6, 1.0, 2.0, 5.0};

    // Methods to compare
    const std::vector<ScreenMethodInfo> methods = {
        {"Screen-TRex Ordinary", ScreenTRexMethod::TREX, false},
        {"Screen-TRex Bootstrap", ScreenTRexMethod::TREX, true},
        {"Screen-TRex Ordinary+DA-AR1", ScreenTRexMethod::TREX_DA_AR1,
            false},
        {"Screen-TRex Bootstrap+DA-AR1", ScreenTRexMethod::TREX_DA_AR1,
            true}
    };

    std::map<std::string, Eigen::VectorXd> est_fdr_map, fdr_map, tpr_map;
    for (const auto& m : methods) {
        const auto len = static_cast<Eigen::Index>(snr_values.size());
        est_fdr_map[m.name] = Eigen::VectorXd::Zero(len);
        fdr_map[m.name]     = Eigen::VectorXd::Zero(len);
        tpr_map[m.name]     = Eigen::VectorXd::Zero(len);
    }

    auto trex_ctrl = make_trex_control();

    // -------------------------------------------------------------------
    // Method loop
    // -------------------------------------------------------------------
    for (const auto& method : methods) {
        std::cout << std::string(70, '=') << "\n";
        std::cout << "Method: " << method.name << "\n";
        std::cout << std::string(70, '=') << "\n";

        auto screen_ctrl = make_screen_control(method.method, method.bootstrap);

        // SNR loop -------------------------------------------------------
        for (std::size_t snr_idx = 0; snr_idx < snr_values.size(); ++snr_idx) {
            const double snr = snr_values[snr_idx];

            const auto make_data = [=](unsigned seed) -> ScrDGPData {
                return make_ar1_dgp(n, p, p1, rho, snr, beta_val, seed);
            };

            const unsigned base_seed = 42u + static_cast<unsigned>(snr_idx) * 1000u;

            std::ostringstream lbl;
            lbl << method.name << "  SNR=" << std::fixed << std::setprecision(2) << snr;
            const auto result = run_mc_trials_screen_trex(
                num_MC, lbl.str(), make_data, trex_ctrl, screen_ctrl, base_seed);

            const auto idx = static_cast<Eigen::Index>(snr_idx);
            est_fdr_map[method.name](idx) = result.avg_est_fdr;
            fdr_map[method.name](idx)     = result.avg_fdr;
            tpr_map[method.name](idx)     = result.avg_tpr;
        }
    }

    std::ostringstream stem;
    stem << "scr_ar1_snr_n" << n << "_p" << p
         << "_rho" << std::fixed << std::setprecision(2) << rho;
    save_and_print_mc_results(num_MC, stem.str(), snr_values, methods,
                              fdr_map, tpr_map, est_fdr_map);
}


// ==============================================================================
// Part 2: SNR sweep — Ordinary vs. Bootstrap vs. DA-EQUI on equicorrelated data
// ==============================================================================

/**
 * @brief SNR-sweep Monte Carlo: Ordinary vs. Bootstrap DA-EQUI Screen-TRex
 *        on equicorrelated data.  Parallelised via OpenMP.
 *
 * @param num_MC    Number of Monte Carlo runs per (method, SNR) pair.
 * @param rho       Equicorrelation coefficient.
 * @param beta_val  Coefficient magnitude for active variables.
 */
void demo_DA_Equi_MonteCarlo(std::size_t num_MC,
                              double     rho      = 0.4,
                              double     beta_val = 1.0)
{
    std::cout.setf(std::ios::unitbuf);
    cdiag::print_section_header(
        "Part 2: DA-EQUI Screen-TRex Monte Carlo on Equicorrelated Data"
    );

    const Eigen::Index n = 300, p = 1000, p1 = 10;
    std::cout << "DGP: Equicorrelated, n=" << n << ", p=" << p << ", p1=" << p1
              << ", rho=" << rho << "\n";

    const std::vector<double> snr_values =
        {0.01, 0.1, 0.2, 0.5, 0.6, 1.0, 2.0, 5.0};

    // Methods to compare
    const std::vector<ScreenMethodInfo> methods = {
        {"Screen-TRex Ordinary", ScreenTRexMethod::TREX, false},
        {"Screen-TRex Bootstrap", ScreenTRexMethod::TREX, true},
        {"Screen-TRex Ordinary DA-EQUI", ScreenTRexMethod::TREX_DA_EQUI,
             false},
        {"Screen-TRex Bootstrap DA-EQUI", ScreenTRexMethod::TREX_DA_EQUI,
            true}
    };

    std::map<std::string, Eigen::VectorXd> est_fdr_map, fdr_map, tpr_map;
    for (const auto& m : methods) {
        const auto len = static_cast<Eigen::Index>(snr_values.size());
        est_fdr_map[m.name] = Eigen::VectorXd::Zero(len);
        fdr_map[m.name]     = Eigen::VectorXd::Zero(len);
        tpr_map[m.name]     = Eigen::VectorXd::Zero(len);
    }

    auto trex_ctrl = make_trex_control();

    // -------------------------------------------------------------------
    // Method loop
    // -------------------------------------------------------------------
    for (const auto& method : methods) {
        std::cout << std::string(70, '=') << "\n";
        std::cout << "Method: " << method.name << "\n";
        std::cout << std::string(70, '=') << "\n";

        auto screen_ctrl = make_screen_control(method.method, method.bootstrap);

        // SNR loop -------------------------------------------------------
        for (std::size_t snr_idx = 0; snr_idx < snr_values.size(); ++snr_idx) {
            const double snr = snr_values[snr_idx];

            const auto make_data = [=](unsigned seed) -> ScrDGPData {
                return make_equi_dgp(n, p, p1, rho, snr, beta_val, seed);
            };

            const unsigned base_seed = 42u + static_cast<unsigned>(snr_idx) * 1000u;

            std::ostringstream lbl;
            lbl << method.name << "  SNR=" << std::fixed << std::setprecision(2) << snr;
            const auto result = run_mc_trials_screen_trex(
                num_MC, lbl.str(), make_data, trex_ctrl, screen_ctrl, base_seed);

            const auto idx = static_cast<Eigen::Index>(snr_idx);
            est_fdr_map[method.name](idx) = result.avg_est_fdr;
            fdr_map[method.name](idx)     = result.avg_fdr;
            tpr_map[method.name](idx)     = result.avg_tpr;
        }
    }

    std::ostringstream stem;
    stem << "scr_equi_snr_n" << n << "_p" << p
         << "_rho" << std::fixed << std::setprecision(2) << rho;
    save_and_print_mc_results(num_MC, stem.str(), snr_values, methods,
                              fdr_map, tpr_map, est_fdr_map);
}


// ==============================================================================
// Part 3: rho sweep — DA-AR1 on AR(1)-correlated data
// ==============================================================================

/**
 * @brief rho-sweep Monte Carlo: Ordinary vs. Bootstrap vs. DA-AR1 Screen-TRex
 *        on AR(1)-correlated data at fixed SNR.  Parallelised via OpenMP.
 *
 * @param num_MC    Number of Monte Carlo runs per (method, rho) pair.
 * @param snr       Fixed signal-to-noise ratio.
 * @param beta_val  Coefficient magnitude for active variables.
 */
void demo_DA_AR1_RhoSweep_MonteCarlo(std::size_t num_MC,
                                     double      snr      = 1.0,
                                     double      beta_val = 1.0)
{
    std::cout.setf(std::ios::unitbuf);
    cdiag::print_section_header(
        "Part 3: DA-AR1 Screen-TRex rho Sweep on AR(1) Data"
    );

    const Eigen::Index n = 300, p = 1000, p1 = 10;
    std::cout << "DGP: AR(1), n=" << n << ", p=" << p << ", p1=" << p1
              << ", SNR=" << snr << "\n";

    const std::vector<double> rho_values =
        {0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9};

    const std::vector<ScreenMethodInfo> methods = {
        {"Screen-TRex Ordinary",         ScreenTRexMethod::TREX,
            false},
        {"Screen-TRex Bootstrap",        ScreenTRexMethod::TREX,
            true },
        {"Screen-TRex Ordinary+DA-AR1",  ScreenTRexMethod::TREX_DA_AR1,
            false},
        {"Screen-TRex Bootstrap+DA-AR1", ScreenTRexMethod::TREX_DA_AR1,
            true }
    };

    std::map<std::string, Eigen::VectorXd> est_fdr_map, fdr_map, tpr_map;
    for (const auto& m : methods) {
        const auto len = static_cast<Eigen::Index>(rho_values.size());
        est_fdr_map[m.name] = Eigen::VectorXd::Zero(len);
        fdr_map[m.name]     = Eigen::VectorXd::Zero(len);
        tpr_map[m.name]     = Eigen::VectorXd::Zero(len);
    }

    auto trex_ctrl = make_trex_control();

    for (const auto& method : methods) {
        std::cout << std::string(70, '=') << "\n";
        std::cout << "Method: " << method.name << "\n";
        std::cout << std::string(70, '=') << "\n";

        auto screen_ctrl = make_screen_control(method.method, method.bootstrap);

        for (std::size_t rho_idx = 0; rho_idx < rho_values.size(); ++rho_idx) {
            const double rho = rho_values[rho_idx];

            const auto make_data = [=](unsigned seed) -> ScrDGPData {
                return make_ar1_dgp(n, p, p1, rho, snr, beta_val, seed);
            };

            const unsigned base_seed = 42u + static_cast<unsigned>(rho_idx) * 1000u;

            std::ostringstream lbl;
            lbl << method.name << "  rho=" << std::fixed << std::setprecision(2) << rho;
            const auto result = run_mc_trials_screen_trex(
                num_MC, lbl.str(), make_data, trex_ctrl, screen_ctrl, base_seed);

            const auto idx = static_cast<Eigen::Index>(rho_idx);
            est_fdr_map[method.name](idx) = result.avg_est_fdr;
            fdr_map[method.name](idx)     = result.avg_fdr;
            tpr_map[method.name](idx)     = result.avg_tpr;
        }
    }

    std::ostringstream stem;
    stem << "scr_ar1_rho_n" << n << "_p" << p
         << "_snr" << std::fixed << std::setprecision(2) << snr;
    save_and_print_mc_results(num_MC, stem.str(), rho_values, methods,
                              fdr_map, tpr_map, est_fdr_map, "rho");
}


// ==============================================================================
// Part 4: rho sweep — DA-EQUI on equicorrelated data
// ==============================================================================

/**
 * @brief rho-sweep Monte Carlo: Ordinary vs. Bootstrap vs. DA-EQUI Screen-TRex
 *        on equicorrelated data at fixed SNR.  Parallelised via OpenMP.
 *
 * @param num_MC    Number of Monte Carlo runs per (method, rho) pair.
 * @param snr       Fixed signal-to-noise ratio.
 * @param beta_val  Coefficient magnitude for active variables.
 */
void demo_DA_Equi_RhoSweep_MonteCarlo(
    std::size_t num_MC,
    double      snr      = 1.0,
    double      beta_val = 1.0)
{
    std::cout.setf(std::ios::unitbuf);
    cdiag::print_section_header(
        "Part 4: DA-EQUI Screen-TRex rho Sweep on Equicorrelated Data"
    );

    const Eigen::Index n = 300, p = 1000, p1 = 10;
    std::cout << "DGP: Equicorrelated, n=" << n << ", p=" << p << ", p1=" << p1
              << ", SNR=" << snr << "\n";

    const std::vector<double> rho_values =
        {0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9};

    const std::vector<ScreenMethodInfo> methods = {
        {"Screen-TRex Ordinary",          ScreenTRexMethod::TREX,
            false},
        {"Screen-TRex Bootstrap",         ScreenTRexMethod::TREX,
            true },
        {"Screen-TRex Ordinary DA-EQUI",  ScreenTRexMethod::TREX_DA_EQUI,
             false},
        {"Screen-TRex Bootstrap DA-EQUI", ScreenTRexMethod::TREX_DA_EQUI,
             true }
    };

    std::map<std::string, Eigen::VectorXd> est_fdr_map, fdr_map, tpr_map;
    for (const auto& m : methods) {
        const auto len = static_cast<Eigen::Index>(rho_values.size());
        est_fdr_map[m.name] = Eigen::VectorXd::Zero(len);
        fdr_map[m.name]     = Eigen::VectorXd::Zero(len);
        tpr_map[m.name]     = Eigen::VectorXd::Zero(len);
    }

    auto trex_ctrl = make_trex_control();

    for (const auto& method : methods) {
        std::cout << std::string(70, '=') << "\n";
        std::cout << "Method: " << method.name << "\n";
        std::cout << std::string(70, '=') << "\n";

        auto screen_ctrl = make_screen_control(method.method, method.bootstrap);

        for (std::size_t rho_idx = 0; rho_idx < rho_values.size(); ++rho_idx) {
            const double rho = rho_values[rho_idx];

            const auto make_data = [=](unsigned seed) -> ScrDGPData {
                return make_equi_dgp(n, p, p1, rho, snr, beta_val, seed);
            };

            const unsigned base_seed = 42u + static_cast<unsigned>(rho_idx) * 1000u;

            std::ostringstream lbl;
            lbl << method.name << "  rho=" << std::fixed << std::setprecision(2) << rho;
            const auto result = run_mc_trials_screen_trex(
                num_MC, lbl.str(), make_data, trex_ctrl, screen_ctrl, base_seed);

            const auto idx = static_cast<Eigen::Index>(rho_idx);
            est_fdr_map[method.name](idx) = result.avg_est_fdr;
            fdr_map[method.name](idx)     = result.avg_fdr;
            tpr_map[method.name](idx)     = result.avg_tpr;
        }
    }

    std::ostringstream stem;
    stem << "scr_equi_rho_n" << n << "_p" << p
         << "_snr" << std::fixed << std::setprecision(2) << snr;
    save_and_print_mc_results(num_MC, stem.str(), rho_values, methods,
                              fdr_map, tpr_map, est_fdr_map, "rho");
}


// ==============================================================================
// Part 5: rho sweep — DA-BLOCK-EQUI on block-equicorrelated data
// ==============================================================================

/**
 * @brief rho-sweep Monte Carlo: Ordinary vs. Bootstrap vs. DA-BLOCK-EQUI
 *        Screen-TRex on block-equicorrelated data at fixed SNR.
 *        Parallelised via OpenMP.
 *
 * @param num_MC    Number of Monte Carlo runs per (method, rho) pair.
 * @param n_blocks  Number of blocks in DGP.
 * @param snr       Fixed signal-to-noise ratio.
 * @param beta_val  Coefficient magnitude for active variables.
 */
void demo_DA_BlockEqui_RhoSweep_MonteCarlo(
    std::size_t num_MC,
    std::size_t n_blocks = 5,
    double      snr      = 1.0,
    double      beta_val = 1.0)
{
    std::cout.setf(std::ios::unitbuf);
    cdiag::print_section_header(
        "Part 5: DA-BLOCK-EQUI Screen-TRex rho Sweep on Block-Equicorr Data"
    );

    const Eigen::Index n = 300, p = 1000, p1 = 10;
    std::cout << "DGP: Block-Equicorr, n=" << n << ", p=" << p << ", p1=" << p1
              << ", n_blocks=" << n_blocks << ", SNR=" << snr << "\n";

    const std::vector<double> rho_values =
        {0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9};

    const std::vector<ScreenMethodInfo> methods = {
        {"Screen-TRex Ordinary",          ScreenTRexMethod::TREX,
            false},
        {"Screen-TRex Bootstrap",         ScreenTRexMethod::TREX,
            true },
        {"Screen-TRex Ord+DA-BLOCK-EQUI", ScreenTRexMethod::TREX_DA_BLOCK_EQUI,
            false},
        {"Screen-TRex Boot+DA-BLOCK-EQUI", ScreenTRexMethod::TREX_DA_BLOCK_EQUI,
            true }
    };

    std::map<std::string, Eigen::VectorXd> est_fdr_map, fdr_map, tpr_map;
    for (const auto& m : methods) {
        const auto len = static_cast<Eigen::Index>(rho_values.size());
        est_fdr_map[m.name] = Eigen::VectorXd::Zero(len);
        fdr_map[m.name]     = Eigen::VectorXd::Zero(len);
        tpr_map[m.name]     = Eigen::VectorXd::Zero(len);
    }

    auto trex_ctrl = make_trex_control();

    for (const auto& method : methods) {
        std::cout << std::string(70, '=') << "\n";
        std::cout << "Method: " << method.name << "\n";
        std::cout << std::string(70, '=') << "\n";

        auto screen_ctrl = make_screen_control(method.method, method.bootstrap,
                                               n_blocks);

        for (std::size_t rho_idx = 0; rho_idx < rho_values.size(); ++rho_idx) {
            const double rho = rho_values[rho_idx];

            const auto make_data = [=](unsigned seed) -> ScrDGPData {
                return make_block_equi_dgp(n, p, p1, rho, n_blocks, snr, beta_val, seed);
            };

            const unsigned base_seed = 42u + static_cast<unsigned>(rho_idx) * 1000u;

            std::ostringstream lbl;
            lbl << method.name << "  rho=" << std::fixed << std::setprecision(2) << rho;
            const auto result = run_mc_trials_screen_trex(
                num_MC, lbl.str(), make_data, trex_ctrl, screen_ctrl, base_seed);

            const auto idx = static_cast<Eigen::Index>(rho_idx);
            est_fdr_map[method.name](idx) = result.avg_est_fdr;
            fdr_map[method.name](idx)     = result.avg_fdr;
            tpr_map[method.name](idx)     = result.avg_tpr;
        }
    }

    std::ostringstream stem;
    stem << "scr_block_equi_rho_n" << n << "_p" << p
         << "_blocks" << n_blocks
         << "_snr" << std::fixed << std::setprecision(2) << snr;
    save_and_print_mc_results(num_MC, stem.str(), rho_values, methods,
                              fdr_map, tpr_map, est_fdr_map, "rho");
}


// ==============================================================================
// Main
// ==============================================================================

int main()
{
    std::cout.setf(std::ios::unitbuf);
    omp_set_num_threads(6);

    // Part 1: SNR sweep — Ordinary vs. Bootstrap vs. DA-AR1 on AR(1) data
    demo_DA_AR1_MonteCarlo(200, 0.5, 1.0);

    // Part 2: SNR sweep — Ordinary vs. Bootstrap vs. DA-EQUI on equicorrelated data
    demo_DA_Equi_MonteCarlo(200, 0.4, 3.0);

    // Part 3: rho sweep — Ordinary vs. Bootstrap vs. DA-AR1 on AR(1) data
    demo_DA_AR1_RhoSweep_MonteCarlo(200, 1.0, 1.0);

    // Part 4: rho sweep — Ordinary vs. Bootstrap vs. DA-EQUI on equicorrelated data
    demo_DA_Equi_RhoSweep_MonteCarlo(200, 1.0, 3.0);

    // Part 5: rho sweep — Ordinary vs. Bootstrap vs. DA-BLOCK-EQUI on block-equicorr data
    demo_DA_BlockEqui_RhoSweep_MonteCarlo(200, 5, 1.0, 3.0);

    return 0;
}

File: Python/trex_selector_methods/trex_screening/demo_trex_scr_03_mc_sim_correlated/demo_trex_scr_03_mc_sim_correlated.py
# ==============================================================================
# demo_trex_scr_03_mc_sim_correlated.py
# ==============================================================================
#
# Screen-TRex Selector, demo 03: correlated designs & dependency-aware screening.
#
# Ordinary screening thresholds each variable's voting proportion Phi_j > 0.5
# independently. Under correlated predictors, redundant correlated variables
# *split* the voting evidence, so Phi drops below 0.5 for all of them and power
# collapses. The dependency-aware (DA) variants pre-group strongly correlated
# variables before voting so the evidence is not split:
#   * TREX_DA_AR1        - for AR(1) (banded) correlation,
#   * TREX_DA_EQUI       - for equicorrelation (single shared factor),
#   * TREX_DA_BLOCK_EQUI - for block-equicorrelation (per-block factors).
#
# Mirrors:
#   cpp/trex_selector_methods/trex_screening/demo_trex_scr_03_mc_sim_correlated/
#       demo_trex_scr_03_mc_sim_correlated.cpp
#
# Demo content (five Monte Carlo studies, matching the cpp Parts 1-5):
#   Part 1: SNR sweep on AR(1) data (rho=0.5): Ordinary/Bootstrap vs. +DA-AR1.
#   Part 2: SNR sweep on equicorrelated data (rho=0.4): vs. +DA-EQUI.
#   Part 3: rho sweep on AR(1) data (SNR=1): vs. +DA-AR1.
#   Part 4: rho sweep on equicorrelated data (SNR=1): vs. +DA-EQUI.
#   Part 5: rho sweep on block-equicorrelated data (SNR=1, 5 blocks): vs. +DA-BLOCK-EQUI.
#
# Downscaled from cpp: the cpp demo uses num_MC = 200; committed default is
# smaller here for a practical Python runtime (override with SCR_NUM_MC). Support
# indices are 0-based.
# ==============================================================================

import os
import sys

_THIS_DIR = os.path.dirname(os.path.abspath(__file__))
_PARENT_DIR = os.path.dirname(_THIS_DIR)
for _p in (_THIS_DIR, _PARENT_DIR):
    if _p not in sys.path:
        sys.path.insert(0, _p)

from trex_scr_common import run_mc_screen, save_and_print_scr_mc

_OUT_DIR = os.path.join(_THIS_DIR, "simulation_results", "data")

# Committed defaults (cpp uses num_MC = 200; downscaled here - five heavy MC parts).
_NUM_MC      = int(os.environ.get("SCR_NUM_MC", 50))
_NUM_WORKERS = int(os.environ.get("SCR_NUM_WORKERS", 6))

N, P, P1 = 300, 1000, 10
SNR_GRID = [0.01, 0.1, 0.2, 0.5, 0.6, 1.0, 2.0, 5.0]
RHO_GRID = [0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9]


def _methods_for(da_method, da_label):
    """Four cpp-style method descriptors: Ordinary/Bootstrap, plain and +DA."""
    return [
        {"name": "Screen-TRex Ordinary: TLARS",              "trex_method": "TREX",     "bootstrap": False},
        {"name": "Screen-TRex Bootstrap-CI: TLARS",          "trex_method": "TREX",     "bootstrap": True},
        {"name": f"Screen-TRex Ordinary+{da_label}: TLARS",     "trex_method": da_method, "bootstrap": False},
        {"name": f"Screen-TRex Bootstrap-CI+{da_label}: TLARS", "trex_method": da_method, "bootstrap": True},
    ]


def run_sweep(part_title, methods, dgp_kind, sweep_name, sweep_values,
              fixed, beta_val, file_stem, screen_n_blocks=None,
              num_MC=_NUM_MC, num_workers=_NUM_WORKERS):
    """Generic MC sweep over SNR or rho for a list of methods (mirrors the cpp parts)."""
    print("\n" + "=" * 70)
    print(part_title)
    print("=" * 70 + "\n")
    method_names = [m["name"] for m in methods]

    def init():
        return {nm: [0.0] * len(sweep_values) for nm in method_names}
    fdr_map, tpr_map, est_fdr_map = init(), init(), init()

    trex_args = {"solver": "TLARS", "use_memory_mapping": False, "K": 20}

    for m in methods:
        print("=" * 70)
        print(f"Method: {m['name']}")
        print("=" * 70)
        screen_args = {"trex_method": m["trex_method"], "bootstrap": m["bootstrap"]}
        if screen_n_blocks is not None:
            screen_args["screen_n_blocks"] = screen_n_blocks
        for vi, v in enumerate(sweep_values):
            snr = v if sweep_name == "SNR" else fixed["snr"]
            rho = v if sweep_name == "rho" else fixed["rho"]
            dgp_args = {"dgp_kind": dgp_kind, "n": N, "p": P, "p1": P1,
                        "snr": snr, "rho": rho, "beta_val": beta_val}
            if dgp_kind == "block_equi":
                dgp_args["dgp_n_blocks"] = fixed["n_blocks"]
            base_seed = 42 + vi * 1000        # mirrors cpp base_seed = 42 + idx * 1000
            res = run_mc_screen(num_MC, f"{m['name']}  {sweep_name}={v:.2f}",
                                dgp_args, trex_args, screen_args,
                                base_seed=base_seed, max_workers=num_workers)
            fdr_map[m["name"]][vi] = res["avg_fdr"]
            tpr_map[m["name"]][vi] = res["avg_tpr"]
            est_fdr_map[m["name"]][vi] = res["avg_est_fdr"]

    save_and_print_scr_mc(num_MC, _OUT_DIR, file_stem, sweep_values,
                          method_names, fdr_map, tpr_map, est_fdr_map,
                          sweep_label=sweep_name)


# ==============================================================================
# Main
# ==============================================================================

def main():
    # Part 1: SNR sweep - AR(1), rho=0.5, beta=1.0
    run_sweep("Part 1: DA-AR1 Screen-TRex Monte Carlo on AR(1) Data",
              _methods_for("TREX_DA_AR1", "DA-AR1"), "ar1", "SNR", SNR_GRID,
              fixed={"rho": 0.5}, beta_val=1.0,
              file_stem="scr_ar1_snr_n300_p1000_rho0.50")

    # Part 2: SNR sweep - equicorrelated, rho=0.4, beta=3.0
    run_sweep("Part 2: DA-EQUI Screen-TRex Monte Carlo on Equicorrelated Data",
              _methods_for("TREX_DA_EQUI", "DA-EQUI"), "equi", "SNR", SNR_GRID,
              fixed={"rho": 0.4}, beta_val=3.0,
              file_stem="scr_equi_snr_n300_p1000_rho0.40")

    # Part 3: rho sweep - AR(1), SNR=1.0, beta=1.0
    run_sweep("Part 3: DA-AR1 Screen-TRex rho Sweep on AR(1) Data",
              _methods_for("TREX_DA_AR1", "DA-AR1"), "ar1", "rho", RHO_GRID,
              fixed={"snr": 1.0}, beta_val=1.0,
              file_stem="scr_ar1_rho_n300_p1000_snr1.00")

    # Part 4: rho sweep - equicorrelated, SNR=1.0, beta=3.0
    run_sweep("Part 4: DA-EQUI Screen-TRex rho Sweep on Equicorrelated Data",
              _methods_for("TREX_DA_EQUI", "DA-EQUI"), "equi", "rho", RHO_GRID,
              fixed={"snr": 1.0}, beta_val=3.0,
              file_stem="scr_equi_rho_n300_p1000_snr1.00")

    # Part 5: rho sweep - block-equicorrelated, SNR=1.0, beta=3.0, 5 blocks
    run_sweep("Part 5: DA-BLOCK-EQUI Screen-TRex rho Sweep on Block-Equicorr Data",
              _methods_for("TREX_DA_BLOCK_EQUI", "DA-BLOCK-EQUI"), "block_equi",
              "rho", RHO_GRID, fixed={"snr": 1.0, "n_blocks": 5}, beta_val=3.0,
              file_stem="scr_block_equi_rho_n300_p1000_blocks5_snr1.00",
              screen_n_blocks=5)

    print("\ndemo_trex_scr_03_mc_sim_correlated.py complete.")


if __name__ == "__main__":
    main()
